Feature / community crud

djlim00 opened the pull request:
## 📚 개요

게시글 CRUD, 댓글 CRUD, 게시글 좋아요 기능을 구현했습니다.
더 좋은 방법이 있거나 JPA관련해서 미흡한 부분이 있을 수 있습니다,,

## 🕐 리뷰 예상 시간

> 10m

## ❗❗ 중요한 변경점(Option)

- 게시물, 댓글, 좋아요 기능을 구현하기 위해서 새로운 테이블과 칼럼이 추가된 부분이 많습니다.

djlim00 committed:
Chore entity오타수정

djlim00 committed:
Move 게시판 관련 엔티티 위치 변경

domain/article/entity로 이동

djlim00 committed:
Feature 게시물 생성, 조회 기능 구현

- 게시물 생성 구현
- 게시판에서 조회 시 게시물 요약본 반환

djlim00 committed:
DB Article, Comment 테이블 구조 변경

- 게시물 CRUD 및 댓글 CRUD를 위해 테이블 구조 변경

djlim00 committed:
Delete Comment.kt

djlim00 committed:
Feat 게시글 상세조회, 수정, 삭제 구현

- 게시물 상세 조회에서 댓글, 이미지까지 가져와야해서 엔티티 수정

djlim00 committed:
Feat 댓글 생성 구현

- 대댓글 기능을 위해서 Comment 엔티티에 부모 commentId 추가

djlim00 committed:
Chore DTO 폴더 위치 변경

djlim00 committed:
Update ArticleService.kt

djlim00 committed:
feat: 작성자 확인을 위해 userId 추가

- 댓글에 작성자 여부를 파악하기 위해서 CommentResponse에 userId 추가

djlim00 committed:
Feat 게시물 좋아요 기능 구현

- article_like 테이블 생성
- ArticleLike Controller, Service, Repository 생성
- 좋아요 버튼을 누르고, 이미 누른 게시물은 취소, 아니라면 좋아요 추가

djlim00 committed:
Feature 댓글생성 시 게시물 ID 파라미터 추가

djlim00 committed:
Update ArticleController.kt

djlim00 committed:
Update 카카오 로그인 확인코드 추가

djlim00 committed:
bugfix: changelog 누락문제 해결

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/controller/ArticleController.kt:
```diff
@@ -0,0 +1,67 @@
+import org.springframework.stereotype.Controller
+import org.springframework.web.bind.annotation.*
+
+@Controller
```
```suggestion
@RestController
```

restcontroller를 사용하시면 기본적으로 모든 return값에 ResponseEntity.ok()를 감싸줍니다.

djlim00 replied:
아 맞네요! 뭔가 안써도 됐던거 같았는데,, 싶었는데 반영했습니다!

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/service/KakaoAuthService.kt:
```diff
@@ -29,6 +29,9 @@ class KakaoAuthService(
         ).execute()
+        println("HTTP Status: ${response.code()}")
+        println("Body: ${response.body()}")
+        println("Error Body: ${response.errorBody()?.string()}")
```
이 print문들은 필요없을 것 같은데 지워주시죠

djlim00 replied:
넵 반영했습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/CommentService.kt:
```diff
@@ -0,0 +1,44 @@
+class CommentService(
+    private val commentRepository: CommentRepository,
+    private val articleRepository: ArticleRepository,
+    private val userRepository: UserRepository
```
service에서는 다른 entity의 repository를 불러오는 것보다 service를 불러오는게 더 좋아보입니다.

djlim00 replied:
ArtivlceService와 CommentService 모두 다른 Entity의 Repository에 의존하지 않고 Service에 의존하도록 반영하였습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/CommentService.kt:
```diff
@@ -0,0 +1,44 @@
+    private val articleRepository: ArticleRepository,
+    private val userRepository: UserRepository
+) {
+    @Transactional
```
트랜잭션 연 이유가 있나요??

djlim00 replied:
article을 가져오고, user가 존재하는지 확인하고, 부모댓글이 있는지 확인하고,  댓글 생성까지 해야되서 댓글을 만드는 과정에서 원자성을 보장하려고 넣었습니다!

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/ArticleService.kt:
```diff
@@ -0,0 +1,126 @@
+
+    fun deleteArticle(articleId: Long) {
+        val article = articleRepository.findById(articleId)
+            .orElseThrow { IllegalArgumentException("해당 게시글이 존재하지 않습니다.") }
```
Exception은 global>exception>CustomException을 throw해주시면 돼고
새로운 CustomError를 선언해서 재사용해주시면 됩니다.

djlim00 replied:
넵 반영했습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/ArticleService.kt:
```diff
@@ -0,0 +1,126 @@
+            duration.toHours() < 24 -> "${duration.toHours()}시간 전"
+            else -> "${duration.toDays()}일 전"
+        }
+    }
```
이런건 오히려 프론트에 맡겨도 되는 부분 아닐까요??

그리고 만약 필요하다고 하더라도 외부에 노출될 필요가 없는 함수는 private으로 설정해주세요

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/ArticleService.kt:
```diff
@@ -0,0 +1,126 @@
+                likeCount = article.likeCount,
+                commentCount = commentCount,
+                uploadAt = uploadAt,
+                thumbnailImage = "이미지",
```
이런 부분에다가는 todo 주석 달아서 나중에 발견하면 고칠 수 있게 해주면 좋습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/ArticleService.kt:
```diff
@@ -0,0 +1,126 @@
+            articleKindId = request.articleKind,
+        )
+
+        val saved = articleRepository.save(article)
```
```suggestion
        val article = articleRepository.save(Article(
            title = request.title,
            text = request.text,
            likeCount = 0,
            isAnonymous = request.isAnonymous,
            userId = request.userId,
            articleKindId = request.articleKind,
        ))
```

이렇게 작성해도 됩니다.

djlim00 replied:
오 반영했습니다!

Ji-InPark reviewed (commented):
일단은 1차로..

djlim00 committed:
Update ArticleController.kt

- @Controller를 @RestController로 변환해서 ResponseEntity 생략

djlim00 committed:
refacfor: 의존성 변경 및 예외처리 수정

- 타 도메인의 repository를 의존하는 방식이 아닌 service단을 의존하도록 변경
- 예외처리를  CustonExeption으로 하도록 변경

djlim00 committed:
bugfix: UserService 중복문제 해결

- UserSerivce를 두개 만들어서 에러가 발생한 부분을 수정했습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/global/exception/CustomError.kt:
```diff
@@ -11,6 +11,12 @@ enum class CustomError(val status: HttpStatus, val message: String) {
+    ARTICLE_NOT_FOUND(NOT_FOUND, "해당 게시글이 존재하지 않습니다."),
+
+    // comment
+    COMMENT_NOT_FOUND(NOT_FOUND, "부모 댓글이 존재하지 않습니다."),
```
```suggestion
    PARENT_COMMENT_NOT_FOUND(NOT_FOUND, "부모 댓글이 존재하지 않습니다."),
```

djlim00 replied:
복붙의 폐해..

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/user/service/UserService.kt:
```diff
@@ -15,4 +17,9 @@ class UserService(
         return userRepository.save(newUser)
     }
+
+    fun getUserOrThrow(userId: Long): User {
```
```suggestion
    fun getUser(userId: Long): User {
```
이렇게만 써도 크게 문제 없는게 어차피 return 타입은 User고 문제가 있을 때 throw하는걸 우리가 따로 try catch하지 않고 프론트까지 넘겨주기 때문에 그렇습니다.

djlim00 replied:
아하 알겠습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/ArticleService.kt:
```diff
@@ -119,8 +121,13 @@ class ArticleService(
         articleRepository.delete(article)
     }
+
+    fun getArticleOrThrow(articleId: Long): Article {
```
```suggestion
    fun getArticle(articleId: Long): Article {
```

djlim00 replied:
반영했습니다~

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/ArticleService.kt:
```diff
@@ -74,10 +77,9 @@ class ArticleService(
     fun getArticleDetail(articleId: Long): ArticleResponse {
         val article = articleRepository.findByIdOrNull(articleId)
-            ?: throw IllegalArgumentException("해당 게시글이 존재하지 않습니다.")
+            ?: throw CustomException(CustomError.ARTICLE_NOT_FOUND)
```
이 부분은 다른 부분과 다르게 findByIdOrNull을 사용하네요. 이유가 있을까요?

djlim00 replied:
아 이유가 없습니다. 그냥 처음에 코틀린의 null처리 대해서 좀 알아보다가 코드들 바꾸면서 처음에 한게 남아있던 것이라서 orElseThrow로 바꿔서 반영하겠습니다!

djlim00 committed:
Update ArticleService.kt

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/ArticleService.kt:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    fun getArticlePreviews(articleKind: ArticleKind?, pageable: Pageable): Page<ArticlePreviewResponse> {
+        val articles: Page<Article> = if (articleKind != null) {
```
```suggestion
        val articles = if (articleKind != null) {
```

djlim00 replied:
쓸데없이 자료형을 썼네용

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/ArticleLikeService.kt:
```diff
@@ -0,0 +1,35 @@
+) {
+    @Transactional
+    fun toggleLike(request: ArticleLikeRequest): ArticleLikeResponse {
+        val article = articleRepository.findById(request.articleId)
+            .orElseThrow { CustomException(CustomError.ARTICLE_NOT_FOUND) }
```
```suggestion
        val article = articleService.getArticle(request.articleId)
```

djlim00 replied:
지금까지와 같은 내용이군요.. 반영하겠습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/controller/CommentController.kt:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.web.bind.annotation.RequestBody
+import org.springframework.web.bind.annotation.RequestMapping
+
+@Controller
```
```suggestion
@RestController
```

djlim00 replied:
반영했습니다~

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/controller/ArticleLikeController.kt:
```diff
@@ -0,0 +1,26 @@
+import org.springframework.web.bind.annotation.RequestBody
+import org.springframework.web.bind.annotation.RequestMapping
+
+@Controller
```
```suggestion
@RestController
```

djlim00 replied:
반영했습니다~

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/controller/ArticleController.kt:
```diff
@@ -0,0 +1,62 @@
+    fun updateArticle(
+        @PathVariable articleId: Long,
+        @RequestBody request: UpdateArticleRequest
+    ): Map<String, String> {
+        articleService.updateArticle(articleId, request)
+        return mapOf("message" to "${articleId}번 게시물이 수정되었습니다.")
```
개인적으로 map 쓰는거 굉장히 안좋아해서요.
위에서했던 것처럼 그냥 response dto 만들어서 내려주는게 더 좋을 것 같아요

djlim00 replied:
UpdateArticleResponse와 DeleteArticleResponse만들어서 반영했습니다~

djlim00 committed:
Update CommentController.kt

- @RestController로 변경

djlim00 committed:
Update ArticleLikeController.kt

- @RestController로 변경

djlim00 committed:
refactor: 예외처리명 수정

- 댓글부모가 없다는 에러를 정확하게 표현되지 않아서 이름 수정

djlim00 committed:
chore: getUser로 함수명 변경

- 기존의 getUserOrThrow에서 getUser로 변경

djlim00 committed:
Update CommentService.kt

- CommentService가 articleRepository가 아니라 articleService에 의존하도록 변경

djlim00 committed:
Update ArticleLikeService.kt

- ArticleLikeService가 ArticleRepository가 아니라 ArticleService에 의존하도록 변경

djlim00 committed:
Update ArticleService.kt

- 게시글 목록 반환하는 방법 간결하게 코드 수정
- 일단 날짜 변환코드 private으로 변경
- 예외처리 방식 ~orNull에서 orElseThrow로 변경
- 함수명 getArticleOrThrow에서 getArticle로 변경

djlim00 committed:
update: 게시물 수정, 삭제 map반환에서 dto반환으로 수정

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/controller/ArticleController.kt:
```diff
@@ -0,0 +1,60 @@
+        return articleService.updateArticle(articleId, request)
+    }
+
+    @Operation(summary = "게시물 삭제", description = "게시물을 삭제하는 API입니다.")
```
KakaoAuthController 참고해서 SwaggerCustomException 달아주세요~

djlim00 replied:
반영했슴둥

Ji-InPark reviewed (commented):
이번 리뷰만 반영하시면 머지 가능할거 같아요~

djlim00 committed:
update: 댓글수정삭제 DTO적용 및 SwaggerCustomExeption적용

- 기존의 Map을 이용해서 응답을 반환하던 Comment 수정, 삭제를 별도의 DTO를 만들어서 수행
- 모든 Controller에 @SwaggerCustomExeption 적용

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/CommentService.kt:
```diff
@@ -39,4 +38,31 @@ class CommentService(
+            .orElseThrow { CustomException(CustomError.COMMENT_NOT_FOUND) }
+
+        comment.text = request.text
+        comment.isAnonymous = request.isAnonymous
```
익명이 바뀌는 경우가 있을까요??

djlim00 replied:
댓글을 최초에 작성할 때는 피그마 보면 익명 여부를 체크로 하는데, 수정할 때도 같은 화면으로 익명여부를 바꿀 수 있다고 보고 만들었는데 이건 어떻게 될지 정해야 할 것 같습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/CommentService.kt:
```diff
@@ -39,4 +38,31 @@ class CommentService(
+        )
+    }
+
+    fun deleteComment(commentId: Long): DeleteCommentResponse {
+        val comment = commentRepository.findById(commentId)
+            .orElseThrow { CustomException(CustomError.COMMENT_NOT_FOUND) }
+
+        commentRepository.delete(comment)
+
+        return DeleteCommentResponse(
+            commentId = comment.id,
+            message = "${commentId}번 댓글이 삭제되었습니다."
+        )
+    }
```
이렇게만 지우면 fk 때문에 제대로 안지워질것 같아요.
부모 댓글 지우려고 한 번 시도해보시죠

djlim00 replied:
아 맞네요. 깜빡하고 바로 바꿔볼게요.

djlim00 replied:
지금 이런식으로 부모 댓글을 지우면 댓글 엔티티에서 
`@OneToMany(mappedBy = "parent", cascade = [CascadeType.ALL], orphanRemoval = true)
    val children: MutableList<Comment> = mutableListOf()`

  이렇게 해놓아서 부모 댓글을 지우면 자식들이 한번에 전부 지워지는데, 이걸 어떻게 해야할지 정해야 할 것 같아요.
  1. 부모 댓글을 지우면 자식 댓글도 없어진다.
  2. 어떤 댓글이든(부모, 자식) 삭제되면 그냥 "삭제된 댓글입니다"로 바뀐다.
  아무래도 2번이 좋겠죠?

Ji-InPark replied:
soft delete 방식이군요.
전 자식도 다 없어지는것도 나쁘지 않다고 생각하는데 안드의 의견을 들어보죠

Ji-InPark reviewed (approved):
![image](https://github.com/user-attachments/assets/2624b39d-4ae9-437a-9f83-d9cd4cbd5c2b)
